### DISCUSSION TITLE
Publicode comme langage de modèle

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#69] [General]
Depuis betagouv/mon-entreprise#602, nous communiquons sur le _langage_ Publicode. Cette approche nous a permis d'entamer des discussions fructueuses visant à ré-organiser le code du moteur, le formaliser, trouver les bonnes abstractions et en définitive le rendre facilement ré-utilisable pour d'autres cas d'usage.

Je pense néanmoins qu'il faut préciser la définition du langage Publicode, car il ne s'agit pas d'une alternative à C ou Python, ce que le seul terme de “langage” peut laisser penser. Pour le HTML on parle souvent de “markup language” (“langage de balisage” en français) et pour le CSS de “stylesheet language” / “langage de feuille de style”.

D'une manière analogue, je propose de caractériser Publicode comme “langage de modèles”. Il fonctionne en décomposant un domaine complexe en une liste de règles élémentaires reliées entre elles. Il permet de créer facilement un modèle sur un sujet et de l'explorer dans une interface interactive.

Concernant mon-entreprise, le fait de parler de “modèle” appelle quelques remarques :
* Nous ne prétendons pas à une correspondance exacte entre la loi et notre modèle de règle, à la différence du langage [Catala](https://catala-lang.org/), en particulier plusieurs organisations des règles sont possibles en publicode, et nous choisissons celle que nous utilisons en fonction de choix éditoriaux (mise en avant d'un dispositif) ou de choix d'implémentation.
  exemple: le barème de majoration des heures supplémentaire est implémenté comme une majoration du temps de travail, et non une majoration de la rémunération des heures, ce qui facilite l'implémentation et est équivalent à la définition législative. D'autres modélisations équivalentes sont possibles
* Il nous est possible d'intégrer un modèle simplifié par rapport à la législation (on accepte de ne pas traiter certains cas particuliers).
* Néanmoins, un modèle peut être aussi précis que nécessaire, notre objectif est une précision au centime près sur le plus grand nombre de situations réelles possible

Le but de cette réflexion est aussi de développer un page d’accueil pour Publicode (actuellement on tombe directement sur un tutoriel d'utilisation sans présentation générale).

**Comment by denismerigoux** [comment]
Intéressant, vu ce que tu dis je pense que PubliCodes ressemble beaucoup à Drools (https://www.drools.org/), mais sur un stack web au lieu de la JVM

**Comment by laem** [comment]
Je ne pense pas qu'on puisse comparer publicodes à des "Business Rules Management System". Avant de commencer j'avais regardé tous ces systèmes et c'est vraiment pas la même chose. A la limite, ça correspond à la toute petite partie "contrôles" de publicodes (renommée depuis si j'ai bien suivi).

J'avais déjà développé l'analogie avec les tableurs et je pense qu'elle tient toujours. Pas dans l'utilisation "organiser un weekend entre potes sur google sheets", mais dans l'utilisation pro où l'on crée un *modèle* de calcul, justement. On le nomme "calcul émissions parc de voiture 2.3.xlsx" et on le partage aux collègues. Publicodes permet de standardiser, versionner et rendre réutilisable facilement ces modèles de calcul, si tant est qu'on a un peu de force de dev en interne (donc très loin de pouvoir remplacer Excel dans la plupart des équipes à ce stade 😄 ). 

Edit : et ne rigolez pas en vous disant "ah c'est juste un truc pour remplacer Excel" : plein d'administrations, plein d'entreprises font reposer de lourds décisions sur Excel. A commencer par l'assemblée nationale et le gouvernement dans l'évaluation des politiques publiques qui décident de l'allocation de milliards d'€.

Une utilisation redoutable de publicodes qui vient en complément à mon avis c'est une simple calculette pour ébaucher quelque chose, calculer des ordres de grandeur et pouvoir nommer les variables pour qu'il soit lisible par la suite. C'est ce que développe Bret Victor ici, si vous ne l'avez pas fait je vous invite vivement à la lecture de ce chapitre de son essai : http://worrydream.com/ClimateChange/#tools

Edit : donc en conclusion, j'aime bien ta proposition Maxime.

**Comment by denismerigoux** [comment]
Je suis moi aussi tout à fait d'accord avec la catégorie "langage de modèle", une catégorie déjà bien fournie : https://en.wikipedia.org/wiki/Modeling_language.

> J'avais déjà développé l'analogie avec les tableurs et je pense qu'elle tient toujours. Pas dans l'utilisation "organiser un weekend entre potes sur google sheets", mais dans l'utilisation pro où l'on crée un modèle de calcul, justement. On le nomme "calcul émissions parc de voiture 2.3.xlsx" et on le partage aux collègues. Publicodes permet de standardiser, versionner et rendre réutilisable facilement ces modèles de calcul, si tant est qu'on a un peu de force de dev en interne (donc très loin de pouvoir remplacer Excel dans la plupart des équipes à ce stade smile ).

En fait plein de langages permettent de faire ça. Basiquement tous les langages très faciles d'accès comme Python, R, et surtout Excel et son VBA qui est effectivement très utilisé de manière industrielle.

Je pense que la force de Publicodes ne vient pas du langage en tant que tel mais bien de la connexion forte avec le monde du Web. En effet, Excel, R, Python et co n'ont pas d'intégration native avec l'environnement Web. C'est peut-être explicable parce que les grosses applications Web datent de la fin des années 2000. 

**Pouvoir écrire son modèle en PubliCodes et en quelque clics produire un site Internet qui permette de l'explorer interactivement est à mon humble avis le point fort et distinctif du langage.**

> Nous ne prétendons pas à une correspondance exacte entre la loi et notre modèle de règle, à la différence du langage Catala, en particulier plusieurs organisations des règles sont possibles en publicode, et nous choisissons celle que nous utilisons en fonction de choix éditoriaux (mise en avant d'un dispositif) ou de choix d'implémentation.

Est-ce que vous avez confronté cette vision avec des juristes ? Pour avoir parlé de ça avec les gens de OpenLaw et ma collègue doctorante en droit, c'est en fait très dur de leur faire accepter que l'on écrit un truc "équivalent". Pour eux le texte de loi est presque sacré, et toute déviation nécessite une analyse juridique (et pas informatique) pour établir la conformité. Notamment la notion de "choix éditorial" fera bondir les juristes :)

> Il nous est possible d'intégrer un modèle simplifié par rapport à la législation (on accepte de ne pas traiter certains cas particuliers).

Effectivement, ça va tout à fait dans le sens du langage d'"ébauche". Python a aussi cette mentalité et c'est un super outil pour ça ; je ne doute pas que PubliCodes soit tout aussi "redoutablement" efficace pour ça. Le risque c'est que l'"ébauche" ne se transforme en produit fini. Par exemple, Dropbox a écrit toute son infrastructure en Python. 10 ans plus tard, ils ont tellement de bugs qu'ils sont en train de réécrire les parties critiques en Rust et qu'ils salarient une équipe d'experts en langages de programmation pour : 
* développer un interpréteur alternatif à Python parce que l'interpréteur de base est trop lent
* développer un vérificateur de types (mypy)

**Comment by mquandalle** [comment]
> Est-ce que vous avez confronté cette vision avec des juristes ? Pour avoir parlé de ça avec les gens de OpenLaw et ma collègue doctorante en droit, c'est en fait très dur de leur faire accepter que l'on écrit un truc "équivalent". Pour eux le texte de loi est presque sacré, et toute déviation nécessite une analyse juridique (et pas informatique) pour établir la conformité. Notamment la notion de "choix éditorial" fera bondir les juristes :)

Même si ça déplaît aux juristes, quand j'ai un problème avec mon propriétaire je consulte la fiche http://service-public.fr et pas le Code de l'Habitation sur http://legifrance.gouv.fr. Service-public est éditorialisé, non exhaustif, et agrège des sources de droits multiples (lois, décrets, règlements, etc.), c'est finalement assez proche de ce que l'on fait. On peut voir Catala comme un Legifrance 2.0 et Monentreprise comme service-public 2.0 :) 

Pour ce qui est de l'équivalence de deux implémentations, je ne pensais qu'à un cas de calcul. Un exemple trivial, pour calculer la décote de l'IR on peut faire `45.25 % × (1717 € – IR)` ou de manière équivalente `777 € – 45.25 % × IR`. Peut être qu'un jour on pourra même prouver qu'une implémentation publicode est conforme à sa définition Catala.

**Comment by denismerigoux** [comment]
J'aime beaucoup l'idée de la dualité sevice-public.fr <-> LegiFrance. Je pense que j'avais du mal à saisir cette complémentarité, je vois beaucoup mieux maintenant ! Merci beaucoup pour ces discussions, j'essaie d'aller au bout des arguments pour saisir exactement l'angle de PubliCodes et comment il s'insère dans les initiatives #RulesAsCode. 

J'espère que vous pourrez réutiliser certaines des idées dans votre comm' pour convaincre un maximum de gens :)

**Comment by mquandalle** [comment]
> et ne rigolez pas en vous disant "ah c'est juste un truc pour remplacer Excel" : plein d'administrations, plein d'entreprises font reposer de lourds décisions sur Excel.

Une discussion sur le sujet https://news.ycombinator.com/item?id=24791017

**Comment by laem** [comment]
T'y as trouvé des choses qui ressemblent à publicodes ?

**Comment by rogerburton** [comment]
En lisant vos réflexions et à partir de ma propre expérience sur l'intégration de la législation (surtout paie et compta) dans des moteurs de règles, j'ai constaté ceci : il y a grosso modo de classes de problèmes : l'identification d'une situation en regard des prescription légales et le calcul qui résulte de cette identification.
le besoin côté utilisateur se décompose de la même manière : comment être guidé dans l'arborescence foisonnante des "cas", et une fois le cas déterminé, comment procéder au calcul en fonction des paramètres applicables au cas. Etant entendu que dans nombre de situation, l'on peut s'arrêter à un niveau x de granularité dans l'arborescence des cas (pas des calculs), suffisant pour opérer le calcul. C'est dans ce sens que le calcul est "d'ébauche". Et selon moi seulement dans ce sens : le calcul est indéniablement correct (au sens de opposable à un tiers) ... nonobstant l'identification complète ou partielle du cas (par exemple dans la paie, telle pétouille issue d'une convention collective, modifiant un taux, une assiette, ajoutant un cotisations, etc.).
Dès lors, l'effort sur la calculabilité (sous réserve de disposer des paramètres en vigueur au moment du calcul, ce qui relève de la veille et non du calculateur : en Belgique, un fichier excel ou xml de notre équivalent Urssaf, l'ONSS, fournit l'ensemble des taux et bases de calcul de la paie en vigueur chaque trimestre, ce qui peut être exporté, en y ajoutant un dictionnaire "user-friendly").
Reste la question de l'identification du cas (en partie reprise en compte dans les étapes avant calcul simulateur de paie version Web de l'URSSAF). Dans publicodes, ça se traduit en liste de "identifiant d'un cas" : oui/non (par exemple "CDD" : "oui") et par les mécanismes "conditions" et "applicable si"
C'est seulement là où publicodes peut présenter des limites. Calculer est assez simple (fastidieux, minutieux, mais simple, en Fr comme en BE) et repose sur finalement très peu de mécanismes arithmétiquement élémentaires. Même si le calcul enchaîne de nombreuses étapes avant de "cracher" le résultat final (la liste organisée des "choses" à payer sur une paie, par exemple).
Donc, conceptuellement, il y a bien deux phases: identifier un cas (c'est à dire le jeu de règles de calcul qui sera appliqué), en partant du plus général pour arriver au plus particulier, et opérer le calcul (au niveau du cas identifié et/ou à chaque niveau dans l'arborescence de choix du cas). Pour l'instant, je suis persuadé que cela relève finalement de "mondes" différents.
D'où je tire, pour le moment, qu'un moteur d'inférence est assez bien adapté à l'identification d'un cas (arriver à identifier un cas d'utilisation au plus fin niveau de granularité possible en fonction des donnés en entrée) notamment parce que je ne suis pas certain que les "cas de paie", par exemple, puissent être aisément représentés dans une arborescence classique (qui serait truffées d'exceptions), là où publicodes gagne haut la main le match sur les règles de calcul applicables à un cas.
En termes de lisibilité, praticabilité, transparence, 
l'utilisateur doit avoir au final deux résumés :
l'identification du cas sur base des données d'entrée (CDD dans le régime des intermittents du spectacle, pour une entreprise disposant de la licence dont le numéro Siret est xxxx, pour un technicien de niveau 4 [localisation administrative parfois nécessaire, cfr Alsace-Moselle], rémunéré au cachet, sans convention d'entreprise particulière, etc. - je dis ici n'importe quoi pour l'exemple)
et le calcul de paie du cas, dont le système expose si besoin est les valeurs par défaut choisies lors des calculs quand la ou les données manquent dans la saisie, ainsi que la date de validité des règles, taux et paramètres utilisés
de cette manière, le dispositif est solide, bien plus que "d'ébauche" : DANS l'hypothèse du cas identifié et avec les valeurs par défaut retenues, le calcul final est correct et complet, il est littéralement "opposable".
désolé de m'épancher ainsi (et peut-être de façon triviale ou déjà obsolète pour vous), mais c'est passionnant. 
Au passage, cela croise un post de LAEM sur son blog je crois quant à la législation comme "code".

**Comment by denismerigoux** [comment]
Bonjour Roger, je rejoins tout à fait ce que vous dites sur le problème de "l'identification du cas". Si l'on cherche à avoir un calcul "opposable", alors il faut être sûr de ne pas avoir manqué une exception qui s'applique au cas présent. Mais plus que le soucis technique, cela suppose une connaissance très fine de la loi qui définit ces "cas" ; travaillez-vous avec des juristes spécialisés dans le droit du travail dans le cadre de votre activité professionnelle ?

**Comment by rogerburton** [comment]
Oui. Mais on est plus avancé en BE que pour la FR. Je vais, POC de mon côté, m'attaquer à expliciter complétement les cas (que nous gérons) avec publicodes (créer les YAML de règles, etc.). Avec une petite difficulté, chez nous on doit faire pas mal de calcul sur des dates et durées dans des périodes, c'est mortel, ça. Je me suis aperçu aussi qu'il y a parfois des divergences entre ce que dit la loi (ou ce qu'elle ne dit pas) et l'interprétation "calculable" qu'en fait l'administration. En BE le fait de la "boîte noire" est infernale : l'ONSS (votre URSSAF) recalcule les données que nous lui envoyons et nous renvoie le résultat de ses calculs : dans certains cas, nous n'avons jamais identifié le pourquoi du comment de son calcul sur telle ou telle cotisations et l'administration elle-même s'avère incapable d'expliciter la règle appliquée effectivement dans ses logiciels. on fait pour plus de 100Milllions d'euros de paie chaque année, ça donne une idée de l'enjeu.
Au passage, dans votre démarche de dissémination de votre projet, essayez à l'occasion de prendre contact avec SMALS (https://www.smals.be/fr/) c'est le développeur IT de l'état belge, et ils ont un département "recherche et expérimentation", qui pourrait être inspiré par et votre démarche (transparence du code) et votre projet publicodes.

**Comment by mquandalle** [comment]
Oui on travaille pas mal sur cette notion d'applicabilité des règles qui est au cœur de l'évaluation d'un publicodes, c'est assez mal documenté pour le moment (https://publi.codes/#applicabilité) mais normalement on doit pouvoir gérer tous les cas de figure.

**Comment by lajarre** [comment]
Bonjour @rogerburton pour vos retours d'expérience. Je note que vous prévoyez de travailler sur un PoC.

J'aimerais en savoir un peu plus sur votre utilisation très concrète de Publicodes: l'avez-vous déjà pris en main, dans quel système prévoyez-vous de l'intégrer, etc. ainsi que prendre vos retours sur le langage et les outils associés. Seriez-vous d'accord pour vous rendre disponible échanger rapidement sur le sujet, et potentiellement faire une session de pair programming ensemble? Dans ce cas je vous propose de reprendre la conversation par email sur contact@mon-entreprise.beta.gouv.fr

**Comment by johangirod** [comment]
> Donc, conceptuellement, il y a bien deux phases: identifier un cas (c'est à dire le jeu de règles de calcul qui sera appliqué), en partant du plus général pour arriver au plus particulier, et opérer le calcul (au niveau du cas identifié et/ou à chaque niveau dans l'arborescence de choix du cas). Pour l'instant, je suis persuadé que cela relève finalement de "mondes" différents.

Je crois que notre approche diverge en ce point là. Il n'y a pas d'un côté les cas, de l'autre les calculs, avec la délimitation des cas qui aboutissent aux calculs différents. 

Ici, on part du calcul pour délimiter les cas. C'est parce qu'on veut calculer le montant des cotisations, et parce que l'on sait que la convention collective a un impacte sur ce calcul que cette donnée va être remontée par le moteur comme "cas à compléter" (que nous nommons "missingVariable"). De même pour toutes les variables 'feuilles' ayant un impact sur la donnée à calculer. 

Cette approche est donc fondamentalement itérative : on calcul en même temps la valeur dans le cas présent, et la liste des cas possibles dérivés du calcul de ce cas présent, et on raffine le calcul jusqu'à ce que toutes les valeurs soient répondue.

Si tous les cas "remontés" pour le calcul d'une variable sont répondus dans la situation, sous reserve que le calcul en lui même implémente bien la loi en totalité (ce que nous ne garantissons pas pour l'instant, à la différence de [Catala](https://catala-lang.org/), alors le calcul est complet et "opposable". 

> Avec une petite difficulté, chez nous on doit faire pas mal de calcul sur des dates et durées dans des périodes, c'est mortel, ça

En effet ! On intègre petit à petit les calculs temporels dans le moteur, ce n'est pas une mince affaire, mais nous savons plus ou moins vers quoi nous voulons tendre (#1007 et betagouv/publicodes#14)

>  Au passage, dans votre démarche de dissémination de votre projet, essayez à l'occasion de prendre contact avec SMALS (https://www.smals.be/fr/) c'est le développeur IT de l'état belge, et ils ont un département "recherche et expérimentation", qui pourrait être inspiré par et votre démarche (transparence du code) et votre projet publicodes.

 C'est noté !
